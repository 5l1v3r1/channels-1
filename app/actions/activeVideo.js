// @flow
import actionTypes from '../constants/actionTypes';
import type { ThunkAction, Dispatch, GetState } from '../constants/typeAliases';
import RatingType from '../dataTypes/ratingType';

export function updateActiveVideo(video: any): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const { id } = getState().activeVideo;
    if (id !== video.id) {
      dispatch({
        type: actionTypes.UPDATE_ACTIVE_VIDEO,
        payload: video
      });
    }
  };
}

export function updateActiveVideoCounts(prevState: RatingType, rating: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    let likeCount = Number(getState().activeVideo.likeCount);
    let dislikeCount = Number(getState().activeVideo.dislikeCount);
    switch (rating) {
      case 'like':
        likeCount = prevState.like ? likeCount - 1 : likeCount + 1;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
        break;
      case 'dislike':
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount + 1;
        break;
      default:
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
    }
    dispatch({
      type: actionTypes.UPDATE_ACTIVE_VIDEO,
      payload: {
        likeCount: String(likeCount),
        dislikeCount: String(dislikeCount)
      }
    });
  };
}
