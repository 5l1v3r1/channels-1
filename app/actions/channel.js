// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch } from '../constants/typeAliases';
import ChannelType from '../dataTypes/channelType';
import { toggleChannelList, updateStatus } from './status';
import { enableSearchBar } from './searchBar';

type ThunkAction = (dispatch: Dispatch) => any;

export function selectChannel(channel: ChannelType): ThunkAction {
  return (dispatch: Dispatch) => {
    const { id, title, thumbnail } = channel;
    dispatch(toggleChannelList());
    dispatch(updateActiveChannel(id));
    dispatch(updateStatus(title, thumbnail));
    dispatch(enableSearchBar(title));
  };
}

export function updateActiveChannel(id: string): Action {
  return {
    type: actionTypes.UPDATE_ACTIVE_CHANNEL,
    payload: id
  };
}
