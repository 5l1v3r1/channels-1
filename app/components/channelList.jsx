// @flow
import React from 'react';
import Channel from './channel';
import Loader from './loader';
import BlankState from './blankState';
import ErrorState from './errorState';
import ChannelType from '../dataTypes/channelType';
import styles from './channelList.scss';

const ChannelList = ({
  channels,
  isFetching,
  error,
  isOpen,
  fetchChannels,
  selectChannel,
  toggleChannelList
}: {
  channels: ChannelType[],
  isFetching: boolean,
  error: boolean,
  isOpen: boolean,
  fetchChannels: () => void,
  selectChannel: (id: string) => void,
  toggleChannelList: () => void
}) => {
  if (isFetching || error) {
    return (
      <div className={isOpen ? `${styles.channelList} ${styles['channelList--isOpen']}` : styles.channelList}>
        <div className={styles.closeTarget} onClick={toggleChannelList} />
        <div className={styles.modal}>
          {isFetching ? (
            <div className={styles.loader}>
              <Loader />
            </div>
          ) : (
            <div className={styles.errorState}>
              <ErrorState
                message={'Error requesting channels.'}
                retry={fetchChannels}
              />
            </div>
          )}
        </div>
      </div>
    );
  }

  return (
    <div className={isOpen ? `${styles.channelList} ${styles['channelList--isOpen']}` : styles.channelList}>
      <div className={styles.closeTarget} onClick={toggleChannelList} />
      <div className={styles.modal}>
        {channels.length ? (
          <ul className={styles.channels}>
            {channels.map(channel =>
              <Channel
                key={channel.id}
                title={channel.title}
                thumbnail={channel.thumbnail}
                videoCount={channel.videoCount}
                subscriberCount={channel.subscriberCount}
                isActive={channel.isActive}
                selectChannel={() => selectChannel(channel.id)}
              />
            )}
          </ul>
        ) : (
          <div className={styles.blankState}>
            <BlankState
              message={'No channels found.'}
            />
          </div>
        )}
      </div>
    </div>
  );
}

export default ChannelList;
