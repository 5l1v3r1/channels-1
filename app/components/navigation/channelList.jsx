// @flow
import React from 'react';
import Channel from './channel';
import Loader from '../loader';
import BlankState from '../blankState';
import ErrorState from '../errorState';
import ChannelType from '../../dataModels/channel';
import styles from './channelList.scss';

const ChannelList = ({
  channels,
  isFetching,
  error,
  isOpen,
  fetchChannels,
  selectChannel,
  toggleVisibility
}: {
  channels: ChannelType[],
  isFetching: boolean,
  error: boolean,
  isOpen: boolean,
  fetchChannels: () => void,
  selectChannel: (id: string) => void,
  toggleVisibility: () => void
}) => {
  if (isFetching || error) {
    return (
      <div className={isOpen ? `${styles.channelList} ${styles['channelList--isOpen']}` : styles.channelList}>
        <div className={styles.closeTarget} onClick={toggleVisibility} />
        <div className={styles.modal}>
          {isFetching ? (
            <div className={styles.loader}>
              <Loader />
            </div>
          ) : (
            <div className={styles.errorState}>
              <ErrorState
                message={'Error requesting channels.'}
                retry={fetchChannels}
              />
            </div>
          )}
        </div>
      </div>
    );
  }

  return (
    <div className={isOpen ? `${styles.channelList} ${styles['channelList--isOpen']}` : styles.channelList}>
      <div className={styles.closeTarget} onClick={toggleVisibility} />
      <div className={styles.modal}>
        {channels.length ? (
          <ul className={styles.channels}>
            {channels.map(channel =>
              <Channel
                key={channel.id}
                title={channel.title}
                thumbnail={channel.thumbnail}
                videoCount={channel.videoCount}
                subscriberCount={channel.subscriberCount}
                isActive={channel.isActive}
                selectChannel={() => selectChannel(channel.id)}
              />
            )}
          </ul>
        ) : (
          <div className={styles.blankState}>
            <BlankState
              message={'No channels found.'}
            />
          </div>
        )}
      </div>
    </div>
  );
}

export default ChannelList;
