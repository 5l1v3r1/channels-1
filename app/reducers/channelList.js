// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, ChannelListState as State } from '../constants/typeAliases';
import Channel from '../dataModels/channel';
import channelReducer from './channel';

const initialState = {
  channels: [],
  isFetching: false,
  error: false,
  isOpen: false
};

export default function channelList(state: State = initialState, { type, payload }: Action): State {
  switch (type) {
    case actionTypes.REQUEST_CHANNELS:
      return {
        ...state,
        isFetching: true,
        error: false
      };
    case actionTypes.RECEIVE_CHANNELS:
      return {
        ...state,
        channels: !payload ? [] : payload.map(json => Channel.from(json)),
        isFetching: false,
        error: false
      };
    case actionTypes.CHANNELS_ERROR:
      return {
        ...state,
        isFetching: false,
        error: true
      };
    case actionTypes.TOGGLE_VISIBILITY:
      return {
        ...state,
        isOpen: !state.isOpen
      };
    case actionTypes.SELECT_CHANNEL:
      return {
        ...state,
        channels: state.channels.map(channel => channelReducer(channel, { type, payload }))
      };
    default:
      return state;
  }
}
