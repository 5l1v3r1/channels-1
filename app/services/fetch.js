// @flow
export const Methods = {
  GET: 'GET',
  POST: 'POST'
};

export function authRequest(uri: string): Promise<any> {
  const [url, body] = uri.split('?');
  const init = {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body
  };
  return fetch(url, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res.json();
    });
}

export function getRequest(uri: string): Promise<any> {
  const init = {
    method: Methods.GET,
    headers: { 'Content-Type': 'application/json' }
  };
  return fetch(uri, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res.json();
    });
}

export function postRequest(uri: string, body?: {}): Promise<any> {
  const init = body ?
  {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  } :
  {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
  };
  return fetch(uri, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res;
    });
}
