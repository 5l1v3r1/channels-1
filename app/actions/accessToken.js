// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch } from '../constants/typeAliases';
import { authRequest } from '../services/fetch';
import { getAccessTokenUri } from '../services/uriGenerator';

type ThunkAction = (dispatch: Dispatch) => any;

export function fetchAccessToken(): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(requestAccessToken());
    const uri = getAccessTokenUri();
    return authRequest(uri)
      .then(json => dispatch(receiveAccessToken(json)))
      .catch(() => dispatch(accessTokenError()));
  };
}

export function requestAccessToken(): Action {
  return {
    type: actionTypes.REQUEST_ACCESS_TOKEN
  };
}

export function receiveAccessToken(json: any): Action {
  return {
    type: actionTypes.RECEIVE_ACCESS_TOKEN,
    payload: json.access_token
  };
}

export function accessTokenError(): Action {
  return {
    type: actionTypes.ACCESS_TOKEN_ERROR
  };
}
