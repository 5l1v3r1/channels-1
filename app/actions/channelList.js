// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch, AccessTokenState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

type GetState = () => { accessToken: AccessTokenState };

type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { accessToken } = getState().accessToken;
    const subscriptions = fetchSubscriptions(accessToken);
    subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(accessToken, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          dispatch(receiveChannels(json));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

function fetchSubscriptions(accessToken: string): Promise<any> {
  const uri = getSubscriptionsUri(accessToken);
  return apiRequest(uri, Methods.GET);
}

function requestChannels(): Action {
  return {
    type: actionTypes.REQUEST_CHANNELS
  };
}

function receiveChannels(json: any): Action {
  return {
    type: actionTypes.RECEIVE_CHANNELS,
    payload: json.items
  };
}

function channelsError(): Action {
  return {
    type: actionTypes.CHANNELS_ERROR
  };
}
