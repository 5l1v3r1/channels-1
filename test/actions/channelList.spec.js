import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as channelListActions from '../../app/actions/channelList';
import * as fetch from '../../app/services/fetch';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('../../app/services/uriGenerator');

describe('channel list actions', () => {

  const accessToken = 'ya29.GlyrBCJQJoIYFzocIunVN-CfjQZMG4oyVuAB6v_x_Z3FRnViyPy_deqRdwSAipQtKc9Nb2RudM9UISwI8SGNXxsJ1t3QHddeCdnoCjsM_vhLa9FlFVqMN_seI7oljg';

  it('creates RECEIVE_CHANNELS when fetching subscriptions has resolved with undefined', () => {
    const store = mockStore({ accessToken });

    fetch.apiRequest = jest.fn(() => Promise.resolve({}));

    return store.dispatch(channelListActions.fetchChannels()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates RECEIVE_CHANNELS when fetching subscriptions has resolved with null', () => {
    const store = mockStore({ accessToken });

    fetch.apiRequest = jest.fn(() => Promise.resolve({ items: null }));

    return store.dispatch(channelListActions.fetchChannels()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates CHANNELS_ERROR when fetching subscriptions has been rejected', () => {
    const store = mockStore({ accessToken });

    fetch.apiRequest = jest.fn(() => Promise.reject());

    return store.dispatch(channelListActions.fetchChannels()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates RECEIVE_CHANNELS when fetching channels has been resolved', () => {
    const store = mockStore({ accessToken });

    fetch.apiRequest = jest.fn()
      .mockReturnValueOnce(Promise.resolve({
        items: [
          {
            snippet: {
              resourceId: {
                channelId: 'UCVTlvUkGslCV_h-nSAId8Sw'
              }
            }
          },
          {
            snippet: {
              resourceId: {
                channelId: 'UCW5YeuERMmlnqo4oq8vwUpg'
              }
            }
          }
        ]
      }))
      .mockReturnValueOnce(Promise.resolve({ items: ['DevTips', 'Fun Fun Function'] }));

    return store.dispatch(channelListActions.fetchChannels()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates CHANNELS_ERROR when fetching channels has been rejected', () => {
    const store = mockStore({ accessToken });

    fetch.apiRequest = jest.fn()
      .mockReturnValueOnce(Promise.resolve({
        items: [
          {
            snippet: {
              resourceId: {
                channelId: 'UCVTlvUkGslCV_h-nSAId8Sw'
              }
            }
          },
          {
            snippet: {
              resourceId: {
                channelId: 'UCW5YeuERMmlnqo4oq8vwUpg'
              }
            }
          }
        ]
      }))
      .mockReturnValueOnce(Promise.reject());

    return store.dispatch(channelListActions.fetchChannels()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('should create an action to request channels', () => {
    expect(channelListActions.requestChannels()).toMatchSnapshot();
  });

  it('should create an action to receive channels', () => {
    const json = { items: ['DevTips', 'Fun Fun Function'] };
    expect(channelListActions.receiveChannels(json)).toMatchSnapshot();
  });

  it('should create an action to handle an error', () => {
    expect(channelListActions.channelsError()).toMatchSnapshot();
  });
});
