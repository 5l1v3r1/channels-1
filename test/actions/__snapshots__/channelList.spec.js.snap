// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`channel list actions creates CHANNELS_ERROR when fetching channels has been rejected 1`] = `
Array [
  Object {
    "type": "REQUEST_CHANNELS",
  },
  Object {
    "type": "CHANNELS_ERROR",
  },
]
`;

exports[`channel list actions creates CHANNELS_ERROR when fetching subscriptions has been rejected 1`] = `
Array [
  Object {
    "type": "REQUEST_CHANNELS",
  },
  Object {
    "type": "CHANNELS_ERROR",
  },
]
`;

exports[`channel list actions creates RECEIVE_CHANNELS when fetching channels has been resolved 1`] = `
Array [
  Object {
    "type": "REQUEST_CHANNELS",
  },
  Object {
    "payload": Array [
      "DevTips",
      "Fun Fun Function",
    ],
    "type": "RECEIVE_CHANNELS",
  },
]
`;

exports[`channel list actions creates RECEIVE_CHANNELS when fetching subscriptions has resolved with null 1`] = `
Array [
  Object {
    "type": "REQUEST_CHANNELS",
  },
  Object {
    "payload": Array [],
    "type": "RECEIVE_CHANNELS",
  },
]
`;

exports[`channel list actions creates RECEIVE_CHANNELS when fetching subscriptions has resolved with undefined 1`] = `
Array [
  Object {
    "type": "REQUEST_CHANNELS",
  },
  Object {
    "payload": Array [],
    "type": "RECEIVE_CHANNELS",
  },
]
`;

exports[`channel list actions should create an action to handle an error 1`] = `
Object {
  "type": "CHANNELS_ERROR",
}
`;

exports[`channel list actions should create an action to receive channels 1`] = `
Object {
  "payload": Object {
    "items": Array [
      "DevTips",
      "Fun Fun Function",
    ],
  },
  "type": "RECEIVE_CHANNELS",
}
`;

exports[`channel list actions should create an action to request channels 1`] = `
Object {
  "type": "REQUEST_CHANNELS",
}
`;
